vir/simd_permute.h

```diff
@@ -0,0 +1,229 @@
+ *                  Matthias Kretz <m.kretz@gsi.de>
+ */
+
+// Implements non-members of P2664
+
+#ifndef VIR_SIMD_PERMUTE_H_
+#define VIR_SIMD_PERMUTE_H_
+
+#include "simd_concepts.h"
+
+#if VIR_HAVE_SIMD_CONCEPTS
+#define VIR_HAVE_SIMD_PERMUTE 1
+
+#include "simd.h"
+#include <bit>
+
+namespace vir
+{
+  namespace detail
+  {
+    template <typename F>
+      concept index_permutation_function_nosize = requires(F const& f)
+      {
+	{ f(std::integral_constant<std::size_t, 0>()) } -> std::integral;
+      };
+
+    template <typename F, std::size_t Size>
+      concept index_permutation_function_size = requires(F const& f)
+      {
+	{ f(std::integral_constant<std::size_t, 0>(), std::integral_constant<std::size_t, Size>()) }
+	  -> std::integral;
+      };
+
+    template <typename F, std::size_t Size>
+      concept index_permutation_function
+	= index_permutation_function_size<F, Size> or index_permutation_function_nosize<F>;
+  }
+
+  constexpr int simd_permute_zero = std::numeric_limits<int>::max();
+
+  constexpr int simd_permute_uninit = simd_permute_zero - 1;
+
+  namespace simd_permutations
+  {
+    struct DuplicateEven
+    {
+      consteval unsigned
+      operator()(unsigned i) const
+      { return i & ~1u; }
+    };
+
+    inline constexpr DuplicateEven duplicate_even {};
+
+    struct DuplicateOdd
+    {
+      consteval unsigned
+      operator()(unsigned i) const
+      { return i | 1u; }
+    };
+
+    inline constexpr DuplicateOdd duplicate_odd {};
+
+    template <unsigned N>
+      struct SwapNeighbors
+      {
+	consteval unsigned
+	operator()(unsigned i, auto size) const
+	{
+	  static_assert(size % (2 * N) == 0,
+			"swap_neighbors<N> permutation requires a multiple of 2N elements");
+	  if (std::has_single_bit(N))
+	    return i ^ N;
+	  else if (i % (2 * N) >= N)
+	    return i - N;
+	  else
+	    return i + N;
+	}
+      };
+
+    template <unsigned N = 1u>
+      inline constexpr SwapNeighbors<N> swap_neighbors {};
+
+    template <int Position>
+      struct Broadcast
+      {
+	consteval int
+	operator()(int) const
+	{ return Position; }
+      };
+
+    template <int Position>
+      inline constexpr Broadcast<Position> broadcast {};
+
+    inline constexpr Broadcast<0> broadcast_first {};
+
+    inline constexpr Broadcast<-1> broadcast_last {};
+
+    struct Reverse
+    {
+      consteval int
+      operator()(int i) const
+      { return -1 - i; }
+    };
+
+    inline constexpr Reverse reverse {};
+
+    template <int Offset>
+      struct Rotate
+      {
+	consteval int
+	operator()(int i, auto size) const
+	{ return (i + Offset) % size(); }
+      };
+
+    template <int Offset>
+      inline constexpr Rotate<Offset> rotate {};
+
+    template <int Offset>
+      struct Shift
+      {
+	consteval int
+	operator()(int i, int size) const
+	{
+	  const int j = i + Offset;
+	  if constexpr (Offset >= 0)
+	    return j >= size ? simd_permute_zero : j;
```

Comment:
```cpp
// in Rotate::operator():
return (i + Offset) % size();
// in Shift::operator():
return j >= size ? simd_permute_zero : j;
```

Can you please explain, why `size` needs a `()` in one case and not in the other?

Reply:
`size` is of type `integral_constant<int, Value>`. It is therefore convertible to `int` and has a call operator that returns `int`. Both are equivalent.
... unless it's an operand to an overloaded operator that deduces the operand as `integral_constant`, in which case it can use its value in a constant expression.

My use of `integral_constant` is not consistent yet. I'm still figuring out what I like best.

Reply:
Ah, actually in line 128 it's an `int size`, so not callable. Just a plain `int`.